python/hail/methods/statgen.py

```diff
@@ -1799,6 +1799,100 @@ def realized_relationship_matrix(call_expr) -> BlockMatrix:
     return (bm.T @ bm) / (bm.n_rows / bm.n_cols)
 
 
+@typecheck(entry_expr=expr_float64)
+def correlation(entry_expr) -> BlockMatrix:
```

Comment:
the name seems too general to be what it does -- correlation between rows of a matrix table

Reply:
people will think of `np.corr` which does something very different. I think if we have `hl.correlation`, it should take two arrays.

Reply:
I've changed the name to `row_correlation`` and added flexibility (which I'm still testing) to `BlockMatrix.write_from_entry_expr` so the new implementation is just:
```
    path = new_temp_file()
    BlockMatrix.write_from_entry_expr(entry_expr, path, mean_impute=True, center=True, normalize=True)
    bm = BlockMatrix.read(path)

    return bm @ bm.T
```
There are only two stages, but they have very different parallelism (the first is the number of block columns and the second is the square of the number of block rows (if not sparsified), so I want users to be able to easily factor it themselves if need be, while also providing this simple interface.

Reply:
This isn't a method on block matrix because users would still have to handle missing-ness en route to block matrix.

And we also don't yet have aggregators on block matrix. I've added sum along one or two axes, implemented via multiplication by vector of ones at block level, so one could do element-wise square and then sum to get squared lengths in order to center/normalize rows or columns on the BlockMatrix side after having mean-imputed on the MatrixTable size.

I'm trying to provide a modular implementation of the main use case that minimizes the number of passes over the data. Let me push what I have and we can discuss further.

Reply:
thanks 👍